servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/AbstractTrafficManagementHttpFilter.java

```diff
@@ -0,0 +1,373 @@
+                "onCancellationTicketTerminal");
+        this.onErrorTicketTerminal = requireNonNull(onErrorTicketTerminal, "onErrorTicketTerminal");
+        this.circuitBreakerPartitionsSupplier = requireNonNull(circuitBreakerPartitionsSupplier,
+                "circuitBreakerPartitionsSupplier");
+        this.observer = requireNonNull(observer, "observer");
+    }
+
+    @Override
+    public HttpExecutionStrategy requiredOffloads() {
+        return offloadNone();
+    }
+
+    // Each filter needs a new capacity & breaker partitions functions.
+    final Function<HttpRequestMetaData, CapacityLimiter> newCapacityPartitions() {
+        return capacityPartitionsSupplier.get();
+    }
+
+    final Function<HttpRequestMetaData, CircuitBreaker> newCircuitBreakerPartitions() {
+        return circuitBreakerPartitionsSupplier.get();
+    }
+
+    Single<StreamingHttpResponse> applyCapacityControl(
+            final Function<HttpRequestMetaData, CapacityLimiter> capacityPartitions,
+            final Function<HttpRequestMetaData, CircuitBreaker> circuitBreakerPartitions,
+            @Nullable final ServerListenContext serverListenContext,
+            final StreamingHttpRequest request,
+            @Nullable final StreamingHttpResponseFactory responseFactory,
+            final Function<StreamingHttpRequest, Single<StreamingHttpResponse>> delegate) {
+        return defer(() -> {
+            final long startTime = nanoTime();
+            final CapacityLimiter partition = capacityPartitions.apply(request);
+            if (partition == null) {
+                observer.onRejectedUnmatchedPartition(request);
+                return rejectWhenNotMatchedCapacityPartition ?
+                        handleLocalCapacityRejection(null, request, responseFactory)
+                                .shareContextOnSubscribe() :
+                        handlePassthrough(delegate, request)
+                                .shareContextOnSubscribe();
+            }
+
+            final CircuitBreaker breaker = circuitBreakerPartitions.apply(request);
+            final ContextMap meta = request.context();
+            final Classification classification = classifier.apply(request);
+            Ticket ticket = partition.tryAcquire(classification, meta);
+
+            if (ticket != null) {
+                ticket = new TrackingDelegatingTicket(ticket, request.hashCode());
+            }
+
+            if (ticket == null) {
+                observer.onRejectedLimit(request, partition.name(), meta, classification);
+                return handleLocalCapacityRejection(serverListenContext, request, responseFactory)
+                        .shareContextOnSubscribe();
+            } else if (breaker != null && !breaker.tryAcquirePermit()) {
+                observer.onRejectedOpenCircuit(request, breaker.name(), meta, classification);
+                // Ignore the acquired ticket if breaker was open.
+                ticket.ignored();
```

Comment:
This is questionable, the rational behind the order was that regardless of whether a breaker allows this through, the limiter has a first say, and still occupies resources (i.e., capacity) regardless. There is an argument to be made for either approach, but so far I haven't seen an impact from this flow. My preference would be to stick to this until we have a use-case that drives a change in behavior.

Reply:
My main thought was that we can minimize number of actions (validation steps) on every request flow if we prioritize one check instead of always running 2 checks. 

I'm agreed that the `ticket` flow has higher value and it should be checked, but calling `circuitBreakerPartitions.apply(request)` doesn't do much if we don't invoke `breaker.tryAcquirePermit()`. So, we can defer `circuitBreakerPartitions` call.

Reply:
This is what I had in mind: https://github.com/apple/servicetalk/pull/2920